### ISSUE TITLE
Fallback to project.json or function.json for unknown environment

### ISSUE THREAD

**Issue by maksymkazakov** [apex/apex#513]
I use three environments in my project: testing, staging and prod.
I've specified three project configs: project.testing.json, project.staging.json and project.prod.json. For the function.json I was not going to introduce any env-related changes and kept that one file. From the documentation I understood that there will be a fallback to default JSON config if there is no corresponding env-related one. I got it wrong as I realized now. So, fine, I can create those env-related function configs, not a problem. The issue is that I need to copy all settings into each project/function config and then support the changes in all of them.

Feature request: Would be great to have either config sections as in INI files for each of existing environments, or have this logic:
- lookup for default config file and using it if found
- lookup for env-related config and merging it with default one, if found

In case of config sections there can be a single project.json file like this:
```
{
  "default": {
    "runtime": "nodejs4.3",
    "role": "arn:aws:...",
    ...
  },
  "testing": {
    "environment": {
      "APP_ENV": "testing",
      "SERVICE_URL": "https://testing.myservice.com"
    }
  }
  "staging": {
    "environment": {
      "APP_ENV": "staging",
      "SERVICE_URL": "https://staging.myservice.com"
    }
  },
  "prod": { ... }
}
```

**Comment by DavidTPate** [comment]
Yeah, I did the exact same things as you when I was going through and getting setup. In my opinion I would love to have the files inherit from one another (with overrides).

**Comment by tj** [comment]
Not against it as long as we have a unified rule for inheriting anything, having special cases could get confusing. Even in this case to me it's still not clear if envs would be merged or not, which is partly why I prefer just having the simple rule of no inheritance.

**Comment by maksymkazakov** [comment]
And again about envs, I've just stuck with need to create separate .tf files for AWS provisioning. Right now there is no difference in file for different envs, so I have to create env-named folders and put symlinks in them pointing to actual Terraform config. Would be great to avoid this too.

**Comment by tj** [comment]
@maksymkazakov what about Terraform modules? That's what I do, works fine, though TF modules need some love, it's reasonably DRY though.

**Comment by maksymkazakov** [comment]
@tj Thanks for hint, I'm not much into Terraform possibilities yet. Might be useful for me later.

**Comment by DavidTPate** [comment]
@maksymkazakov Yeah as @tj mentioned Terraform modules make it easy to re-use things. That's how I do it for dealing with my different environments. All you have to do is put the `.tf` files in another directory and then reference it as a module. you can pass in Variables to control what it does and tags and such. 

Really easy to do if you already have your `.tf` files created. Check this for an example of how simple it is: https://www.terraform.io/docs/modules/usage.html#multiple-instances-of-a-module

Keep in mind though that it will change your resource names, so you'll need to re-create them.

**Comment by hsienchiaolee** [comment]
+1 on some inheritance for configuration as well. Most of my environments has the same `project.json` setup, especially through the use of `nameTemplate` (aware of deprecation discussion), creating duplicated `project.env.json` isn't ideal.
